fe/fe-core/src/main/java/com/starrocks/load/pipe/Pipe.java

```diff
@@ -176,6 +189,62 @@ public void schedule() {
         finalizeTasks();
     }
 
+    /**
+     * Recovery after restart, to guarantee exactly-once ingestion
+     * 1. Persist the insert label along with file before loading state
+     * 2. Check the insert success or not by insert-label after restart recovery
+     * 3. As a result, we could clearly distinguish insert success or not,
+     * so that each file would not be ingested repeatedly
+     */
+    public void recovery() {
+        if (recovered) {
+            return;
+        }
+        LOG.info("pipe {} start to recover", name);
+
+        GlobalTransactionMgr txnMgr = GlobalStateMgr.getCurrentGlobalTransactionMgr();
+        long dbId = getPipeId().getDbId();
+        List<PipeFileRecord> loadingFiles =
+                pipeSource.getFileListRepo().listFilesByState(FileListRepo.PipeFileState.LOADING);
+
+        if (CollectionUtils.isEmpty(loadingFiles)) {
+            recovered = true;
+            return;
+        }
+
+        for (PipeFileRecord file : loadingFiles) {
+            if (StringUtils.isEmpty(file.insertLabel)) {
+                file.loadState = FileListRepo.PipeFileState.ERROR;
+            } else {
+                TransactionStatus txnStatus = txnMgr.getLabelStatus(dbId, file.insertLabel);
+                if (txnStatus == null || txnStatus.isFailed()) {
+                    file.loadState = FileListRepo.PipeFileState.ERROR;
+                } else {
+                    file.loadState = FileListRepo.PipeFileState.LOADED;
+                }
+            }
+        }
+
+        List<PipeFileRecord> failedFiles = loadingFiles.stream()
+                .filter(x -> x.loadState == FileListRepo.PipeFileState.ERROR)
+                .collect(Collectors.toList());
+        List<PipeFileRecord> loadedFiles = loadingFiles.stream()
+                .filter(x -> x.loadState == FileListRepo.PipeFileState.LOADED)
+                .collect(Collectors.toList());
+
+        if (CollectionUtils.isNotEmpty(failedFiles)) {
+            pipeSource.getFileListRepo().updateFileState(failedFiles, FileListRepo.PipeFileState.ERROR, null);
+        }
+        if (CollectionUtils.isNotEmpty(loadedFiles)) {
+            pipeSource.getFileListRepo().updateFileState(loadedFiles, FileListRepo.PipeFileState.LOADED, null);
```

Comment:
If you already use the `insert` to replace the `load` command, you should not use `LOADED` to indicate the state. Maybe better to be like a normal state name, like `RUNNING`, or `FINISHED`

Reply:
changed to `FINISHED`

Reply:
BTW, it will be a brand-new functionality, i don't think we need to pay so much attention to the compatibility issue

Reply:
Not for compatibility, just for consistency. The state for ingestion/schema change/materialized view can be use the same state diagram.